runtime/binding-tls/src/main/java/io/aklivity/zilla/runtime/binding/tls/internal/TlsEventContext.java

```diff
@@ -0,0 +1,96 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.aklivity.zilla.runtime.binding.tls.internal;
+
+import java.nio.ByteBuffer;
+import java.util.function.LongSupplier;
+
+import javax.net.ssl.SSLHandshakeException;
+import javax.net.ssl.SSLKeyException;
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.net.ssl.SSLProtocolException;
+
+import org.agrona.MutableDirectBuffer;
+import org.agrona.concurrent.UnsafeBuffer;
+
+import io.aklivity.zilla.runtime.binding.tls.internal.types.event.TlsError;
+import io.aklivity.zilla.runtime.binding.tls.internal.types.event.TlsEventFW;
+import io.aklivity.zilla.runtime.engine.EngineContext;
+import io.aklivity.zilla.runtime.engine.binding.function.MessageConsumer;
+
+public class TlsEventContext
+{
+    private static final int EVENT_BUFFER_CAPACITY = 1024;
+
+    private final TlsEventFW.Builder tlsEventRW = new TlsEventFW.Builder();
+    private final MutableDirectBuffer eventBuffer = new UnsafeBuffer(ByteBuffer.allocate(EVENT_BUFFER_CAPACITY));
+    private final int typeId;
+    private final MessageConsumer logEvent;
+    private final LongSupplier timestamp;
+
+    public TlsEventContext(
+        int typeId,
+        EngineContext context)
+    {
+        this.typeId = typeId;
+        this.logEvent = context.logEvent();
+        this.timestamp = context.timestamp();
+    }
+
+    public void tlsFailed(
+        TlsError error,
+        long traceId)
+    {
+        TlsEventFW event = tlsEventRW
+            .wrap(eventBuffer, 0, eventBuffer.capacity())
+            .tlsFailed(e -> e
+                .timestamp(timestamp.getAsLong())
+                .traceId(traceId)
+                .error(r -> r.set(error))
+            )
+            .build();
+        System.out.println(event); // TODO: Ati
+        logEvent.accept(typeId, event.buffer(), event.offset(), event.limit());
+    }
+
+    public void tlsFailed(
+        Exception ex,
+        long traceId)
+    {
+        TlsError error;
+        if (ex instanceof SSLProtocolException)
+        {
+            error = TlsError.PROTOCOL_ERROR;
+        }
+        else if (ex instanceof SSLKeyException)
+        {
+            error = TlsError.KEY_ERROR;
+        }
+        else if (ex instanceof SSLHandshakeException)
+        {
+            error = TlsError.HANDSHAKE_ERROR;
+        }
+        else if (ex instanceof SSLPeerUnverifiedException)
+        {
+            error = TlsError.PEER_UNVERIFIED_ERROR;
+        }
+        else
+        {
+            error = TlsError.UNSPECIFIED_ERROR;
+        }
```

Comment:
Not sure about these `instanceof` checks, but it may be too verbose to put specific catch blocks in each area where we currently have `catch (SSLException ex)`.

Perhaps there's a way to wrap those calls to add a common try/catch block around them?
Not urgent, so let's discuss this a bit more after the rest of the task is done.

Reply:
Sure. I played around with this for a bit, and I understand this piece of code does not look very elegant, but I didn't find a better solution. I assume we don't want to pollute the eight call sites with the bloated exception subtype checks and the code duplication that it comes with.